File: Day 16/solution.cpp
#include <filesystem>
#include <sstream>
#include <fstream>
#include <iostream>
#include <vector>
#include <utility>
#include <stack>
#include <unordered_map>
#include <unordered_set>
#include <numeric>
#include <algorithm>
#include <execution>
#include <cassert>

constexpr auto max_size = std::numeric_limits<std::streamsize>::max();

typedef std::pair<int, int> coordinate;

// Gotten from: https://stackoverflow.com/a/16473277
struct pair_hash
{
	template <typename T, typename U>
	std::size_t operator()(const std::pair<T, U> &x) const
	{
		auto hash_combine = [](size_t &seed, T const &v)
		{
			seed ^= std::hash<T>()(v) + 0x9e3779b9 + (seed << 6) + (seed >> 2);
		};
		size_t result = std::hash<T>()(x.first);
		hash_combine(result, x.second);
		return result;
	}
};

enum symbols
{
	SPACE = '.',
	POSITIVE_MIRROR = '/',
	NEGATIVE_MIRROR = '\\',
	HORIZONTAL_SPLITTER = '-',
	VERTICAL_SPLITTER = '|'
};

class Solver
{
	std::filesystem::path filepath;

	std::vector<std::string> get_grid()
	{
		std::vector<std::string> result;
		std::ifstream file(this->filepath);
		std::string line;
		while (std::getline(file, line))
		{
			result.push_back(line);
		}
		file.close();
		return result;
	}

	int get_num_energized_positions(const std::vector<std::string> &grid, int i, int j, int d_i, int d_j)
	{
		auto add_next = [&grid = std::as_const(grid)](
												std::stack<std::tuple<int, int, int, int>> &stack,
												int i, int j, int d_i, int d_j)
		{
			switch (grid.at(i).at(j))
			{
			case SPACE:
				stack.emplace(i + d_i, j + d_j, d_i, d_j);
				break;
			case POSITIVE_MIRROR:
				stack.emplace(i - d_j, j - d_i, -d_j, -d_i);
				break;
			case NEGATIVE_MIRROR:
				stack.emplace(i + d_j, j + d_i, d_j, d_i);
				break;
			case VERTICAL_SPLITTER:
				if (d_i != 0)
				{
					stack.emplace(i + d_i, j + d_j, d_i, d_j);
					break;
				}
				stack.emplace(i + 1, j, 1, 0);
				stack.emplace(i - 1, j, -1, 0);
				break;
			case HORIZONTAL_SPLITTER:
				if (d_j != 0)
				{
					stack.emplace(i + d_i, j + d_j, d_i, d_j);
					break;
				}

				stack.emplace(i, j + 1, 0, 1);
				stack.emplace(i, j - 1, 0, -1);
				break;
			default:
				break;
			}
		};

		std::unordered_map<coordinate, std::unordered_set<coordinate, pair_hash>, pair_hash> energized;
		std::stack<std::tuple<int, int, int, int>> stack;
		stack.emplace(i, j, d_i, d_j);
		while (!stack.empty())
		{
			auto [i, j, d_i, d_j] = stack.top();
			stack.pop();

			if (0 > i || i >= grid.size() || 0 > j || j >= grid.front().size())
				continue;

			coordinate position{i, j}, direction{d_i, d_j};
			if (energized[position].contains(direction))
				continue;

			energized[position].insert(direction);
			add_next(stack, i, j, d_i, d_j);
		}

		return energized.size();
	}

	int part_1()
	{
		std::vector<std::string> grid = this->get_grid();
		return this->get_num_energized_positions(grid, 0, 0, 0, 1);
	}

	int part_2()
	{
		std::vector<std::string> grid = this->get_grid();

		int result = 0;
		std::mutex m;

		std::vector<int> range(grid.size());
		std::iota(range.begin(), range.end(), 0);
		std::for_each(
				std::execution::par, range.begin(), range.end(), [this, &grid, &result, &m](int i)
				{
					int current = std::max(
						this->get_num_energized_positions(grid, i, 0, 0, 1), this->get_num_energized_positions(grid, i, grid[i].size() - 1, 0, -1));
					std::lock_guard<std::mutex> guard(m);
					result = std::max(result, current); });

		range = std::vector<int>(grid.front().size());
		std::iota(range.begin(), range.end(), 0);
		std::for_each(
				std::execution::par, range.begin(), range.end(), [this, &grid, &result, &m](int j)
				{
					int current = std::max(this->get_num_energized_positions(grid, 0, j, 1, 0),
					this->get_num_energized_positions(grid, grid.size() - 1, j, -1, 0));
					std::lock_guard<std::mutex> guard(m);
					result = std::max(result, current); });
		return result;
	}

public:
	Solver(std::filesystem::path filepath = "input.txt") : filepath(filepath) {}

	void solve()
	{
		std::cout << "Day 16" << std::endl;
		std::cout << "Solving: " << this->filepath << std::endl;
		std::cout << "Part 1:" << std::endl;
		std::cout << this->part_1() << std::endl;
		std::cout << "---" << std::endl;
		std::cout << "Part 2:" << std::endl;
		std::cout << this->part_2() << std::endl;
	}
};

int main(int argc, char *argv[])
{
	std::unique_ptr<Solver> solver;
	if (argc == 1)
	{
		solver = std::make_unique<Solver>();
	}
	else if (argc == 2)
	{
		solver = std::make_unique<Solver>(argv[1]);
	}
	else if (argc == 3 && (argv[1] == "-p" || argv[1] == "--path"))
	{
		solver = std::make_unique<Solver>(argv[2]);
	}
	else
	{
		return 1;
	}

	solver->solve();
}

File: Day 16/solution.py
"""
Day 16 solution.
"""
import collections
import enum
import sys
from multiprocessing import Pool
from typing import Generator


class Solver:
    """Day 16 solver."""

    class symbols(enum.Enum):
        """Valid input symbols."""

        SPACE = "."
        POSITIVE_MIRROR = "/"
        NEGATIVE_MIRROR = "\\"
        HORIZONTAL_SPLITTER = "-"
        VERTICAL_SPLITTER = "|"

    def __init__(self, filepath: str = "input.txt"):
        self.filepath = filepath

    def get_grid(self) -> list[str]:
        """Parse the grid."""
        with open(
            self.filepath, "r", encoding=sys.getdefaultencoding()
        ) as file:
            return [line.strip() for line in file]

    @staticmethod
    def multiprocessing_get_num_energized_positions_wrapper(
        args: tuple[list[str], int, int, int, int]
    ) -> int:
        """Wrapper to unpack arguments for get_num_energized_positions."""
        return Solver.get_num_energized_positions(*args)

    @staticmethod
    def get_num_energized_positions(
        grid: list[str], i: int, j: int, d_i: int, d_j: int
    ) -> int:
        """Get the number of energized positions."""

        def add_next(
            stack: list[tuple[int, int, int, int]],
            i: int,
            j: int,
            d_i: int,
            d_j: int,
        ):
            c = grid[i][j]
            if c == Solver.symbols.SPACE.value:
                stack.append((i + d_i, j + d_j, d_i, d_j))
            elif c == Solver.symbols.POSITIVE_MIRROR.value:
                stack.append((i - d_j, j - d_i, -d_j, -d_i))
            elif c == Solver.symbols.NEGATIVE_MIRROR.value:
                stack.append((i + d_j, j + d_i, d_j, d_i))
            elif c == Solver.symbols.VERTICAL_SPLITTER.value:
                if d_i != 0:
                    stack.append((i + d_i, j + d_j, d_i, d_j))
                    return

                stack.append((i + 1, j, 1, 0))
                stack.append((i - 1, j, -1, 0))
            elif c == Solver.symbols.HORIZONTAL_SPLITTER.value:
                if d_j != 0:
                    stack.append((i + d_i, j + d_j, d_i, d_j))
                    return

                stack.append((i, j + 1, 0, 1))
                stack.append((i, j - 1, 0, -1))

        energized = collections.defaultdict(set)
        stack = [(i, j, d_i, d_j)]
        while stack:
            i, j, d_i, d_j = stack.pop()

            if not 0 <= i < len(grid) or not 0 <= j < len(grid[0]):
                continue

            if (d_i, d_j) in energized[i, j]:
                continue

            energized[i, j].add((d_i, d_j))
            add_next(stack, i, j, d_i, d_j)

        return len(energized)

    def part_1(self) -> int:
        """Part 1 solver."""
        grid = self.get_grid()
        return self.get_num_energized_positions(grid, 0, 0, 0, 1)

    def part_2(self) -> int:
        """Part 2 solver."""
        grid = self.get_grid()

        def get_start_positions() -> (
            Generator[tuple[list[str], int, int, int, int], None, None]
        ):
            # Start from the left column and go right
            for i in range(len(grid)):
                yield grid, i, 0, 0, 1

            # Start from the right column and go left
            for i, row in enumerate(grid):
                yield grid, i, len(row) - 1, 0, -1

            # Start from the top row and go down
            for j in range(len(grid[0])):
                yield grid, 0, j, 1, 0

            # Start from the bottom row and go up
            for j in range(len(grid[-1])):
                yield grid, len(grid) - 1, j, -1, 0

        with Pool() as pool:
            return max(
                pool.imap(
                    self.multiprocessing_get_num_energized_positions_wrapper,
                    get_start_positions(),
                )
            )

    def solve(self) -> None:
        """Runs part 1 and part 2."""
        print("Day 16")
        print(f"Solving: {self.filepath}")
        print("Part 1:")
        print(self.part_1())
        print("---")
        print("Part 2:")
        print(self.part_2())


if __name__ == "__main__":
    match sys.argv:
        case [_]:
            solver = Solver()
        case [_, "--path", path] | [_, "-p", path] | [_, path]:
            solver = Solver(path)
        case _:
            print("Usage: python solution.py <path>")
            sys.exit(0)
    solver.solve()
